include/mbedtls/kw.h

```diff
@@ -0,0 +1,163 @@
+ * according to <em>NIST SP 800-38F: Recommendation for Block Cipher Modes of Operation:
+ * Methods for Key Wrapping</em>.
+ * Its purpose is to protect cryptographic keys.
+ * It uses units of 8 Bytes called semiblocks. The minimal number of input semiblocks is:
+ * <ul><li>For KW mode: 2 semiblocks.</li>
+ * <li>For KWP mode: 1 semiblock.</li></ul>
+ *
+ */
+/*
+ *  Copyright (C) 2018, Arm Limited (or its affiliates), All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of Mbed TLS (https://tls.mbed.org)
+ */
+
+#ifndef MBEDTLS_KW_H
+#define MBEDTLS_KW_H
+
+#include "cipher.h"
+
+#define MBEDTLS_ERR_KW_BAD_INPUT       -0x0080 /**< Bad input parameters to the function. */
+#define MBEDTLS_ERR_KW_AUTH_FAILED     -0x0082 /**< Authenticated decryption failed. */
+#define MBEDTLS_ERR_KW_HW_ACCEL_FAILED -0x0084 /**< KW hardware accelerator failed. */
+
+#define MBEDTLS_KEY_WRAPPING_MODE_KW    0
+#define MBEDTLS_KEY_WRAPPING_MODE_KWP   1
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*! The 64-bit default ICV for KW mode. */
+#define MBEDTLS_KW_ICV1             {0xA6A6A6A6, 0xA6A6A6A6}
+/*! The 32-bit default ICV for KWP mode. */
+#define MBEDTLS_KW_ICV2             {0xA65959A6}
+
+#if !defined(MBEDTLS_KW_ALT)
+// Regular implementation
+//
+
+/**
+ * \brief    The KW context-type definition. The KW context is passed
+ *           to the APIs called.
+ */
+typedef struct {
+    int                      mode;          /*!< The KW mode used. */
+    mbedtls_cipher_context_t cipher_ctx;    /*!< The cipher context used. */
+}
+mbedtls_kw_context;
+
+#else  /* MBEDTLS_KW_ALT */
+#include "kw_alt.h"
+#endif /* MBEDTLS_KW_ALT */
+
+/**
+ * \brief           This function initializes the specified KW context,
+ *                  to make references valid, and prepare the context
+ *                  for mbedtls_kw_setkey() or mbedtls_kw_free().
+ *
+ * \param ctx       The KW context to initialize.
+ * \param mode      The KW mode to use (MBEDTLS_KEY_WRAPPING_MODE_KW or MBEDTLS_KEY_WRAPPING_MODE_KWP)
+ *
+ */
+void mbedtls_kw_init( mbedtls_kw_context *ctx, int mode );
+
+/**
+ * \brief           This function initializes the KW context set in the
+ *                  \p ctx parameter and sets the encryption key.
+ *
+ * \param ctx       The KW context.
+ * \param cipher    The 128-bit block cipher to use. Currently supports only AES.
```

Comment:
Being compliant to a NIST standard doesn't prevent supporting non-NIST-approved algorithm. NIST does not forbid having support for other algorithms in your source code. You can even certify a product according to FIPS 140 up to level 2 even if it includes support for other algorithms, as long as it doesn't require these algorithms to function and you document that “if [exhaustive list of non-NIST-approved cryptographic algorithms supported by the product] is used then the product is no longer operating in compliant mode”.

I won't block the PR because of this, but I disapprove artificially restricting it to AES. It should support any block cipher with a block size of 16.